marlowe-playground-client/src/JavascriptEditor/State.purs

```diff
@@ -3,35 +3,66 @@ module JavascriptEditor.State where
 import Prelude hiding (div)
 import Data.Array as Array
 import Data.Either (Either(..))
-import Data.Lens (assign, to, use, view)
+import Data.Foldable (for_)
+import Data.Lens (_Just, assign, to, use, view)
+import Data.Lens.Extra (peruse)
 import Data.List ((:))
 import Data.List as List
 import Data.Map as Map
-import Data.Maybe (Maybe(..))
+import Data.Maybe (Maybe(..), fromMaybe)
+import Data.String (drop, joinWith, length, take)
 import Effect.Aff.Class (class MonadAff, liftAff)
 import Halogen (HalogenM, liftEffect, query)
 import Halogen.Blockly as Blockly
 import Halogen.Monaco (Message(..), Query(..)) as Monaco
-import JavascriptEditor.Types (Action(..), CompilationState(..), State, _compilationResult, _keybindings, _showBottomPanel)
+import JavascriptEditor.Types (Action(..), CompilationState(..), State, _compilationResult, _decorationIds, _keybindings, _showBottomPanel)
 import Language.Javascript.Interpreter (_result)
 import Language.Javascript.Interpreter as JSI
 import LocalStorage as LocalStorage
 import MainFrame.Types (ChildSlots, _blocklySlot, _jsEditorSlot)
 import Marlowe (SPParams_)
-import Monaco (isError)
+import Monaco (IRange, isError)
 import Servant.PureScript.Settings (SPSettings_)
 import StaticData (jsBufferLocalStorageKey)
 import StaticData as StaticData
+import Text.Parsing.StringParser.Basic (lines)
+
+checkDecorationPosition :: Int -> Maybe IRange -> Maybe IRange -> Boolean
+checkDecorationPosition numLines (Just { endLineNumber }) (Just { startLineNumber }) = (endLineNumber == headerLength) && (startLineNumber == numLines - footerLength + 1)
+  where
+  headerLength = Array.length decorationHeader
+
+  footerLength = Array.length decorationFooter
+
+checkDecorationPosition _ _ _ = false
 
 handleAction ::
   forall m.
   MonadAff m =>
   SPSettings_ SPParams_ ->
   Action ->
   HalogenM State Action ChildSlots Void m Unit
-handleAction _ (HandleEditorMessage (Monaco.TextChanged text)) = do
-  liftEffect $ LocalStorage.setItem jsBufferLocalStorageKey text
-  assign _compilationResult NotCompiled
+handleAction _ (HandleEditorMessage (Monaco.TextChanged text)) =
+  ( do
+      let
+        prunedText = pruneJSboilerplate text
+
+        numLines = Array.length $ lines text
+      mDecorIds <- peruse (_decorationIds <<< _Just)
+      case mDecorIds of
+        Just decorIds -> do
+          mRangeHeader <- query _jsEditorSlot unit (Monaco.GetDecorationRange decorIds.topDecorationId identity)
+          mRangeFooter <- query _jsEditorSlot unit (Monaco.GetDecorationRange decorIds.bottomDecorationId identity)
+          mContent <- liftEffect $ LocalStorage.getItem jsBufferLocalStorageKey
+          if (mContent == Just prunedText) || (checkJSboilerplate text && checkDecorationPosition numLines mRangeHeader mRangeFooter) then
+            ( do
+                liftEffect $ LocalStorage.setItem jsBufferLocalStorageKey prunedText
+                assign _compilationResult NotCompiled
+            )
```

Comment:
I don't quite understand the logic of the `mContent == Just prunedText` case, doesn't this just set local storage to be what it was previously? Maybe this should be split so that we see the 3 different cases separately and the `mContent == Just prunedText` only sets `_compilationResult`?

Reply:
Now that you mention it, I am not sure it is necessary at all. But is probably good to prevent it from looping if we later introduce a bug by accident. I will label it